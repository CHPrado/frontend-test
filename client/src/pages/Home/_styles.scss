#page-home {
  height: 100vh;

  .content {
    width: 100%;
    height: 100%;
    margin: 0 auto;
    padding: 0 30px;
  
    @extend .flex;
    justify-content: flex-start;
    flex-direction: column;

    header {
      margin: 40px auto 20px;

      @extend .flex;
      flex-direction: column;

      h1 {
        font-size: 35px;
        color: $title-color;
      }

      a {
        width: 100%;
        max-width: 360px;
        height: 50px;
        border-radius: 8px;
        margin-top: 40px;
        text-decoration: none;
        background: $primary-color;

        @extend .flex;

        strong {
          flex: 1;
          text-align: center;
          color: #FFF;
        }
      }

      a:hover {
        background: $button-hover-color;
      }
    }

    main {
      flex: 1;

      @extend .flex;
      flex-direction: column;

      ul {
        display: grid;
        grid-template-columns: repeat(3, 1fr);
        gap: 16px;
        list-style: none;

        li {
          background: #FFF;
          border: 2px solid #FFF;
          border-radius: 8px;
          padding: 32px 24px 16px;
        
          @extend .flex;
          flex-direction: column;
          text-align: center;
        
          img {
            border-radius: 8px;
            width: 300px;
          }

          span {
            margin-top: 12px;
            color: $text-color;
            @extend .flex;
          }

          button {
            margin-top: 12px;
            width: 100%;
            height: 50px;
            background: transparent;
            border-radius: 8px;
            transition: background-color 0.2s;
            border: solid $text-color;

            cursor: pointer;

            svg {
              color: $text-color;
            }
          }

          button:hover {
            border: solid #e00e0e;

            svg {
              color: #e00e0e;
            }
          }
        }
      }
    }
  }
}

@media(max-width: 900px) {
  #page-home .content {
    text-align: center;

    header {
      margin: 48px auto 20px;
    }

    main ul {
      display: grid;
      grid-template-columns: repeat(1, 1fr);
      gap: 16px;
      list-style: none;
    }
  }
}