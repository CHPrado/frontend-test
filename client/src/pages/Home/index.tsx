import React, { useEffect, useState } from 'react';
import { Link } from 'react-router-dom';

import api from '../../services/api';

import { UserProps } from '../../interfaces';
import UserCard from '../../components/UserCard';
import Snack, { SnackProps } from '../../components/Snack';
import AlertBox from '../../components/AlertBox';

const Home = () => {
  const [users, setUsers] = useState<UserProps[]>([]);

  const [alertState, setAlertState] = useState(false);
  const [alertType, setAlertType] = useState<SnackProps['type']>('');
  const [alertMessage, setAlertMessage] = useState('');
  const [alertTitle, setAlertTitle] = useState('');
  const [deleteId, setDeletId] = useState<UserProps['id']>();

  const [snackState, setSnackState] = useState(false);
  const [snackType, setSnackType] = useState<SnackProps['type']>('');
  const [snackMessage, setSnackMessage] = useState('');


  useEffect(() => {
    api.get('users').then((response) => {
      setUsers(response.data);
    });
  }, []);

  const handleDeleteClick = (user: UserProps) => {
    setAlertTitle('Excluir Usuário');
    setAlertMessage(`Deseja excluir o usuário ${user.name}?`);
    setAlertType('warn');
    setAlertState(true);

    setDeletId(user.id);
  };

  const handleAlertConfirm = () => {
    if (typeof deleteId === 'number') handleDeleteUser(deleteId);
  }

  const handleCloseAlert = () => {
    setDeletId(undefined);
    setAlertState(false);
  };

  const handleCloseSnack = () => {
    setSnackState(false);
  };


  async function handleDeleteUser(userId: UserProps['id']) {
    await api.delete(`users/${userId}`).then((response) => {
      setSnackMessage('Usuário excluído!');
      setSnackType('success');
      setSnackState(true);

      setUsers(users.filter((user: UserProps) => user.id !== userId));
    });
  }

  return (
    <>
      <AlertBox 
        type={alertType}
        open={alertState}
        onConfirm={handleAlertConfirm}
        onClose={handleCloseAlert}
        title={alertTitle}
        message={alertMessage}
      />
      <Snack
        type={snackType}
        open={snackState}
        close={handleCloseSnack}
        message={snackMessage}
      />
      <div id="page-home">
        <div className="content">
          <header>
            <h1>
              Lista de Usuários
            </h1>

            <Link to="/create-user">
              <strong>Cadastre um novo usuário</strong>
            </Link>
          </header>

          <main>
            <ul>
              {users.map((user) => <UserCard key={user.id} user={user} handleDeleteClick={() => handleDeleteClick(user)} />)}
            </ul>
          </main>
        </div>
      </div>
    </>
  )
};

export default Home;
