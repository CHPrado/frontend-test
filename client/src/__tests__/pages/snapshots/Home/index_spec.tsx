import React from 'react';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';

import Home from '../../../../pages/Home';

test('it matches the snapshot', () => {
  const tree = renderer.create(
    <MemoryRouter>
      <Home />
    </MemoryRouter>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
