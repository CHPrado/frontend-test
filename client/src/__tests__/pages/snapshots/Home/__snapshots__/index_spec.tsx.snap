// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it matches the snapshot 1`] = `
Array [
  <div
    className="alert-container hide"
  >
    <div
      className="alert-box"
    >
      <div
        className="title-container"
      >
        <strong>
          
        </strong>
        <div
          className="alert-close"
          onClick={[Function]}
          role="button"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
            />
          </svg>
        </div>
      </div>
      <div
        className="message-container"
      >
        <div
          className="alert-icon "
        />
        <div
          className="alert-message"
        >
          <strong>
            
          </strong>
        </div>
      </div>
      <div
        className="buttons-container"
      >
        <button
          onClick={[Function]}
          type="button"
        >
          Cancelar
        </button>
        <button
          className="button-confirm"
          onClick={[Function]}
          type="button"
        >
          Confirmar
        </button>
      </div>
    </div>
  </div>,
  <div
    className="snack-container hide"
  >
    <div
      className="snack-box "
    >
      <div
        className="snack-icon"
      />
      <div
        className="snack-message"
      >
        <strong>
          
        </strong>
      </div>
      <div
        className="snack-close "
        onClick={[Function]}
        role="button"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
          />
        </svg>
      </div>
    </div>
  </div>,
  <div
    id="page-home"
  >
    <div
      className="content"
    >
      <header>
        <h1>
          Lista de Usuários
        </h1>
        <a
          href="/create-user"
          onClick={[Function]}
        >
          <strong>
            Cadastre um novo usuário
          </strong>
        </a>
      </header>
      <main>
        <ul />
      </main>
    </div>
  </div>,
]
`;
